### DISCUSSION TITLE
Can Thingsboard allow communication on other MQTT topics?

### DISCUSSION THREAD

**Discussion by timothyparez** [thingsboard/thingsboard#11851] [Q&A]
_Disclaimer: In this post I am making assumptions based on my limited Thingsboard knowledge,   
so please forgive me ignorance and correct me where I am wrong._

We have a series of clients and devices that can communicate RPC style over various transports,
one of them is MQTT (others are TCP, ZMQ, RS232, BLE, etc...). 
Basically this means we can connect the clients and devices to ANY MQTT
broker (Mosquitto, HiveMQ, etc...) and they are instantly able to communicate with each other.
The broker does not and should not care what the encoding of the requests/responses is (Protocol Buffers in our case).

It looks like this:

![image](https://github.com/user-attachments/assets/cf2abb8c-61cb-4ea6-8c45-4a0fceabd0de)

Specifically we use two MQTT topics, one for requests and another for responses:

- `req/{clientid}/{target}/{deviceid}/{messageid}`
- `res/{deviceid}/{clientid}/{messageid}`

> These design decisions are not important and beyond the topic of this post, 
> we could be talking about any MQTT topic not used by Thingsboard itself.

Since Thingsboard supports MQTT our idea was to simply connect the clients and devices to it,
enabling communication between clients and devices and then add support for Thingsboard  
specific features, like publishing telemetry to the `v1/devices/me/telemetry` topic in JSON format.

 - Connecting to the Thingsboard MQTT service is not a problem, clients and devices connected instantly.
 - Publishing telemetry data to the `v1/devices/me/telemetry` is also not a problem
 - Communication between client and devices however seems to be blocked by Thingsboard.

So now I have some questions:

**Is there an option somewhere to allow communication on other topics not used by Thingsboard?**

If there is, can someone point me to it ?

**If not, why? What's the idea behind this restriction?** 

If the built-in broker can/would allow communication on other topics without caring about
the data encoding, it would be much easier to add Thingsboard support to existing solutions.

Adding a second broker without these limitations just to enable communication is well... silly
The same for another service such as the IoT-Gateway. Our devices already have the ability
to communicate via standard MQTT brokers, why force the introduction of extra complexity?

Side notes:

 - We understand that without an IoT-Gateway Thingsboard cannot analyze
   our custom protocol, but that's ok, we just need it to exchange messages
   like a normal MQTT broker.
 - Our protocol messages are Protocol Buffer v2 encoded, but this should
   not matter either, since we don't need Thingsboard to understand it.
 - Right now our devices are capable of publishing telemetry data to Thingsboard on)
 - We want to avoid setting up a second broker or additional services.

**Comment by K0I05** [comment]
I haven't tested this, but the topics appear to be editable in ThingsBoard under Device Profile.

![Screenshot 2025-05-28 192512](https://github.com/user-attachments/assets/a33ef091-a012-4e97-9332-8ff24601ec9e)

**Comment by vbinupx** [comment]
Hello, have you found a solution? I use configuration settings to customize MQTT themes, and the device has already received `v1/devices/me/test metrics`

**Comment by timothyparez** [comment]
Yes, we ended up simply implementing the Thingsboard conventions as another option in our device.
And then we add support to our device for it to receive RPC commands from Thingsboard in which our own protocol requests and responses are passed as json (which is what thingsboard uses in the RPC requests) and let our device convert that to Protocol buffers internally.

Now our device can connect to a Thingsboard MQTT broker and another normal MQTT broker at the same time,
allowing the customer to use whatever they want.

Other than that, there's not much you can do in Thingsboard itself to allow other topics (afaik)

**Reply by vbinupx** [reply]
Okay, thank you. It seems that I can only use RPC to handle it
